pilot/platform/cloudfoundry/servicediscovery.go

```diff
@@ -0,0 +1,134 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudfoundry
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+
+	copilotapi "code.cloudfoundry.org/copilot/api"
+
+	"istio.io/istio/pilot/model"
+)
+
+type CopilotClient interface {
+	copilotapi.IstioCopilotClient
+}
+
+const CF_APP_PORT = 8080
+
+type ServiceDiscovery struct {
+	Client CopilotClient
+}
+
+// Services list declarations of all services in the system
+func (sd *ServiceDiscovery) Services() ([]*model.Service, error) {
+	resp, err := sd.Client.Routes(context.Background(), new(copilotapi.RoutesRequest))
+	if err != nil {
+		return nil, fmt.Errorf("getting services: %s", err)
+	}
+	services := []*model.Service{}
+
+	for hostname, _ := range resp.Backends {
+		services = append(services, newService(hostname))
+	}
+
+	return services, nil
+}
+
+// GetService retrieves a service by host name if it exists
+func (sd *ServiceDiscovery) GetService(hostname string) (*model.Service, error) {
+	services, err := sd.Services()
+	if err != nil {
+		return nil, err
+	}
+	for _, svc := range services {
+		if svc.Hostname == hostname {
+			return svc, nil
+		}
+	}
+	return nil, fmt.Errorf("No service exists with name '%s'.", hostname)
+}
+
+func newService(hostname string) *model.Service {
+	return &model.Service{
+		Hostname: hostname,
+		Ports: []*model.Port{
+			&model.Port{
+				Port:     CF_APP_PORT,
+				Protocol: model.ProtocolTCP,
+			},
+		},
+	}
+}
+
+// Instances retrieves instances for a service and its ports that match
+// any of the supplied labels. All instances match an empty tag list.
+//
+// For example, consider the example of catalog.mystore.com as described in NetworkEndpoints
+// Instances(catalog.myservice.com, 80) ->
+//      --> NetworkEndpoint(172.16.0.1:8888), Service(catalog.myservice.com), Labels(foo=bar)
+//      --> NetworkEndpoint(172.16.0.2:8888), Service(catalog.myservice.com), Labels(foo=bar)
+//      --> NetworkEndpoint(172.16.0.3:8888), Service(catalog.myservice.com), Labels(kitty=cat)
+//      --> NetworkEndpoint(172.16.0.4:8888), Service(catalog.myservice.com), Labels(kitty=cat)
+//
+// Calling Instances with specific labels returns a trimmed list.
+// e.g., Instances(catalog.myservice.com, 80, foo=bar) ->
+//      --> NetworkEndpoint(172.16.0.1:8888), Service(catalog.myservice.com), Labels(foo=bar)
+//      --> NetworkEndpoint(172.16.0.2:8888), Service(catalog.myservice.com), Labels(foo=bar)
+//
+// Similar concepts apply for calling this function with a specific
+// port, hostname and labels.
+func (sd *ServiceDiscovery) Instances(hostname string, ports []string,
+	tagsList model.LabelsCollection) ([]*model.ServiceInstance, error) {
+	resp, err := sd.Client.Routes(context.Background(), new(copilotapi.RoutesRequest))
+	if err != nil {
+		return nil, fmt.Errorf("getting instances: %s", err)
+	}
+	service := newService(hostname)
+	instances := []*model.ServiceInstance{}
```

Comment:
we also know the length of the `instances` array here based on `backendSet.GetBackends()` so we can use `make([]T, len, cap)` here as above